src/include/txn_inline.h

```diff
@@ -968,8 +970,12 @@ __wt_txn_read(
         return (0);
     WT_ASSERT(session, cbt->upd_value->type == WT_UPDATE_INVALID);
 
-    /* If there is no ondisk value, there can't be anything in the history store either. */
-    if (cbt->ref->page->dsk == NULL || cbt->slot == UINT32_MAX) {
+    /*
+     * If there is no ondisk value, there can't be anything in the history store either. For a row
+     * store, we can't have a history store value if the key is found in the insert list.
+     */
+    if (cbt->ref->page->dsk == NULL || cbt->slot == UINT32_MAX ||
```

Comment:
Do we still need this?

Reply:
Yes. If there is no on-disk value, three is no point in searching the datastore is the same for both row and column stores. What extra is for the row-store, where it is possible that there is an on-disk page but the key may not exist?

Reply:
I think column store doesn't even use slot so this check doesn't apply. For row store, the second check is enough? For column store, I think we need to check whether the cell is deleted or not.

Reply:
if (cbt->ref->page->dsk == NULL || (S2BT(session)->type == BTREE_ROW && cbt->ins != NULL) || (S2BT(session)->type == COLUMN || CELL_IS_DELETED) something like this?

Reply:
I checked all the code flows and nowhere we called the `__wt_txn_read` when the `cbt->slot` is `UINT32_MAX`. Removing that condition doesn't affect. But to add the condition that you specified, we need to access the cell. Anyway, later the cell is accessed and verified. So I feel there is no need to add an extra check here.

Reply:
OK

Reply:
I don't find any code that checks the cell is deleted. What is the time window returned for deleted cell? The stop ts will be 0 so it is always visible?

Reply:
The CELL_DEL are read with a time window of stop as 0 during the unpack, and it generates a tombstone and returns to the application.